File: bff/bff_service.py
import os
import logging
from typing import Any, Dict, Optional
import httpx

from bff.models import (
    BffRiskCalculationRequest,
    BffFinancialEngineRequest,
    BffPolePriorityRequest,
    BffEnrichedRiskResponse,
    BffEnrichedFinancialResponse,
    BffEnrichedPriorityResponse,
    BffUnifiedAuditResponse,
)

logger = logging.getLogger("bff.service")

BACKEND_SERVICE_URL = os.getenv("BACKEND_SERVICE_URL", "http://localhost:8000")


class BffService:
    def __init__(self, backend_url: str = BACKEND_SERVICE_URL):
        self.backend_url = backend_url.rstrip("/")

    async def calculate_risk(self, req: BffRiskCalculationRequest) -> BffEnrichedRiskResponse:
        """Call backend Risk Calculation API and enrich response for React UI."""
        payload = {
            "asset_id": req.asset_id,
            "tenant_id": req.tenant_id,
            "ahs": {
                "age_years": req.age_years,
                "material": req.material,
                "remaining_fiber_pct": req.remaining_fiber_pct,
                "defect_severity": req.defect_severity,
                "lean_deg": req.lean_deg,
                "attachment_count": req.attachment_count,
                "asset_class": req.asset_class,
                "reinforced_within_10_years": req.reinforced_within_10_years,
            },
            "consequence": {
                "population_exposure": req.population_exposure,
                "critical_infrastructure": req.critical_infrastructure,
                "service_impact": req.service_impact,
            },
            "cqs": {
                "failure_cost": req.failure_cost,
                "outage_duration_hours": req.outage_duration_hours,
                "customer_count": req.customer_count,
            },
            "weather": {
                "wind_speed_mph": req.wind_speed_mph,
                "temperature_f": req.temperature_f,
                "relative_humidity_pct": req.relative_humidity_pct,
                "precipitation_in": req.precipitation_in,
                "fire_weather_index": req.fire_weather_index,
            },
            "days_overdue": req.days_overdue,
            "work_already_scheduled": req.work_already_scheduled,
        }

        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.post(f"{self.backend_url}/api/v1/risks/calculate", json=payload)
            resp.raise_for_status()
            data = resp.json()

        score = data.get("risk_score", 0.0)
        category = data.get("risk_category", "UNKNOWN")

        # Color mapping & action recommendation
        color_map = {
            "CRITICAL": "#ef4444",
            "VERY_HIGH": "#f97316",
            "HIGH": "#eab308",
            "MODERATE": "#3b82f6",
            "LOW": "#10b981",
        }
        color = color_map.get(category.upper(), "#6b7280")

        if category in ("CRITICAL", "VERY_HIGH"):
            rec = "Immediate field dispatch and structural reinforcement required."
        elif category == "HIGH":
            rec = "Schedule maintenance work within 30 days. Priority queue escalated."
        elif category == "MODERATE":
            rec = "Monitor during next routine cycle. Routine inspection recommended."
        else:
            rec = "Asset operating within normal safety limits. No immediate action."

        factor_scores = {
            "AHS": data.get("ahs", 0.0),
            "CES": data.get("ces", 0.0),
            "CQS": data.get("cqs", 0.0),
            "DHM": data.get("dhm", 0.0),
            "OPS": data.get("ops", 0.0),
            "VES": data.get("ves", 0.0),
        }

        return BffEnrichedRiskResponse(
            risk_evaluation_id=data.get("risk_evaluation_id", ""),
            asset_id=data.get("asset_id", req.asset_id),
            tenant_id=data.get("tenant_id", req.tenant_id),
            risk_score=score,
            risk_category=category,
            category_color=color,
            action_recommendation=rec,
            confidence=data.get("confidence", 1.0),
            calculated_at=data.get("calculated_at", ""),
            factor_scores=factor_scores,
            drivers=data.get("explanation", {}).get("factors", []),
            warnings=data.get("warnings", []),
        )

    async def evaluate_financials(self, req: BffFinancialEngineRequest) -> BffEnrichedFinancialResponse:
        """Call backend Financial Engine API and format results."""
        payload = {
            "asset_id": req.asset_id,
            "tenant_id": req.tenant_id,
            "condition_band": req.condition_band,
            "replacement_cost": req.replacement_cost,
            "customers": req.customers,
            "outage_hours": req.outage_hours,
            "outage_cost_per_customer_hour": req.outage_cost_per_customer_hour,
            "wildfire_liability_exposure": req.wildfire_liability_exposure,
            "hftd_tier": req.hftd_tier,
            "action_cost": req.action_cost,
            "remaining_life": req.remaining_life,
            "design_life": req.design_life,
            "discount_rate": req.discount_rate,
            "degradation_rate": req.degradation_rate,
        }

        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.post(f"{self.backend_url}/api/v1/financial-engine/evaluate", json=payload)
            resp.raise_for_status()
            data = resp.json()

        base_summary = data.get("base_summary", {})
        net_benefit = base_summary.get("net_benefit", 0.0)
        roi = base_summary.get("roi", 0.0)
        eal = base_summary.get("eal", 0.0)

        fmt_net_benefit = f"${net_benefit:,.2f}"
        fmt_roi = f"{roi * 100:.1f}%"
        fmt_eal = f"${eal:,.2f}"

        if roi > 2.0:
            invest_rec = "Strong positive return on investment. Immediate capital allocation justified."
        elif roi > 0.0:
            invest_rec = "Favorable net benefit. Recommended for upcoming budget allocation."
        else:
            invest_rec = "Negative net benefit. Defer major overhaul unless required by risk threshold."

        return BffEnrichedFinancialResponse(
            evaluation_id=data.get("evaluation_id", ""),
            asset_id=data.get("asset_id", req.asset_id),
            tenant_id=data.get("tenant_id", req.tenant_id),
            calculation_version=data.get("calculation_version", "1.0"),
            scenarios=data.get("scenarios", {}),
            base_summary=base_summary,
            investment_recommendation=invest_rec,
            formatted_net_benefit=fmt_net_benefit,
            formatted_roi=fmt_roi,
            formatted_eal=fmt_eal,
        )

    async def calculate_pole_priority(self, req: BffPolePriorityRequest) -> BffEnrichedPriorityResponse:
        """Call backend Pole Priority API and compute work queue tier."""
        payload = {
            "asset_id": req.asset_id,
            "tenant_id": req.tenant_id,
            "composite_risk_score": req.composite_risk_score,
            "consequence_score": req.consequence_score,
            "hazard_exposure_score": req.hazard_exposure_score,
            "days_overdue": req.days_overdue,
            "inspection_confidence": req.inspection_confidence,
        }

        async with httpx.AsyncClient(timeout=15.0) as client:
            resp = await client.post(f"{self.backend_url}/api/v1/risks/pole-priority", json=payload)
            resp.raise_for_status()
            data = resp.json()

        priority_score = data.get("priority_score", 0.0)
        if priority_score >= 80.0:
            tier = "P1 - EMERGENCY / IMMEDIATE"
            color = "#ef4444"
            urgency = "Dispatch field crew immediately. High public safety risk."
        elif priority_score >= 60.0:
            tier = "P2 - HIGH URGENCY"
            color = "#f97316"
            urgency = "Queue for work order authorization within 7 days."
        elif priority_score >= 40.0:
            tier = "P3 - MEDIUM PRIORITY"
            color = "#eab308"
            urgency = "Schedule replacement in current quarterly plan."
        else:
            tier = "P4 - ROUTINE / DEFERRABLE"
            color = "#10b981"
            urgency = "No work order required at present."

        return BffEnrichedPriorityResponse(
            asset_id=req.asset_id,
            tenant_id=req.tenant_id,
            priority_score=priority_score,
            priority_tier=tier,
            priority_color=color,
            urgency_recommendation=urgency,
            score_breakdown=data.get("breakdown", {}),
        )

    async def run_unified_audit(self, asset_id: str, tenant_id: str) -> BffUnifiedAuditResponse:
        """Orchestrate risk, financial evaluation, priority, and history into a single payload."""
        risk_req = BffRiskCalculationRequest(asset_id=asset_id, tenant_id=tenant_id)
        fin_req = BffFinancialEngineRequest(asset_id=asset_id, tenant_id=tenant_id)

        risk_res = await self.calculate_risk(risk_req)
        fin_res = await self.evaluate_financials(fin_req)

        prio_req = BffPolePriorityRequest(
            asset_id=asset_id,
            tenant_id=tenant_id,
            composite_risk_score=risk_res.risk_score,
            consequence_score=risk_res.factor_scores.get("CES", 50.0),
            hazard_exposure_score=risk_res.factor_scores.get("DHM", 50.0),
        )
        prio_res = await self.calculate_pole_priority(prio_req)

        history_count = 0
        async with httpx.AsyncClient(timeout=10.0) as client:
            try:
                hist_resp = await client.get(f"{self.backend_url}/api/v1/risks/history/{asset_id}?tenant_id={tenant_id}")
                if hist_resp.status_code == 200:
                    history_count = len(hist_resp.json())
            except Exception as e:
                logger.warning(f"Could not fetch history for asset {asset_id}: {e}")

        health_status = "STABLE"
        if risk_res.risk_category in ("CRITICAL", "VERY_HIGH") or prio_res.priority_score >= 80:
            health_status = "CRITICAL ACTION REQUIRED"
        elif risk_res.risk_category == "HIGH":
            health_status = "ELEVATED RISK"

        from datetime import datetime, timezone
        return BffUnifiedAuditResponse(
            asset_id=asset_id,
            tenant_id=tenant_id,
            timestamp=datetime.now(timezone.utc).isoformat(),
            risk=risk_res,
            financials=fin_res,
            priority=prio_res,
            history_count=history_count,
            overall_health_status=health_status,
        )

File: dotnet-bff/Services/BffService.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Nodes;
using DotNetBff.Models;

namespace DotNetBff.Services;

public class BffService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<BffService> _logger;

    public BffService(HttpClient httpClient, ILogger<BffService> logger)
    {
        _httpClient = httpClient;
        _logger = logger;
    }

    public async Task<string> CheckBackendHealthAsync()
    {
        try
        {
            var response = await _httpClient.GetAsync("/health");
            if (response.IsSuccessStatusCode)
            {
                return "healthy";
            }
            // Try fallback / health endpoint
            var responseAlt = await _httpClient.GetAsync("/");
            return responseAlt.IsSuccessStatusCode ? "healthy" : "unreachable";
        }
        catch (Exception ex)
        {
            _logger.LogWarning("Backend health check failed: {Message}", ex.Message);
            return "unreachable";
        }
    }

    public async Task<BffEnrichedRiskResponse> CalculateRiskAsync(BffRiskCalculationRequest req)
    {
        var payload = new
        {
            asset_id = req.AssetId,
            tenant_id = req.TenantId,
            ahs = new
            {
                age_years = req.AgeYears,
                material = req.Material,
                remaining_fiber_pct = req.RemainingFiberPct,
                defect_severity = req.DefectSeverity,
                lean_deg = req.LeanDeg,
                attachment_count = req.AttachmentCount,
                asset_class = req.AssetClass,
                reinforced_within_10_years = req.ReinforcedWithin10Years
            },
            consequence = new
            {
                population_exposure = req.PopulationExposure,
                critical_infrastructure = req.CriticalInfrastructure,
                service_impact = req.ServiceImpact
            },
            cqs = new
            {
                failure_cost = req.FailureCost,
                outage_duration_hours = req.OutageDurationHours,
                customer_count = req.CustomerCount
            },
            weather = new
            {
                wind_speed_mph = req.WindSpeedMph,
                temperature_f = req.TemperatureF,
                relative_humidity_pct = req.RelativeHumidityPct,
                precipitation_in = req.PrecipitationIn,
                fire_weather_index = req.FireWeatherIndex
            },
            days_overdue = req.DaysOverdue,
            work_already_scheduled = req.WorkAlreadyScheduled
        };

        JsonObject? backendData = null;
        try
        {
            var response = await _httpClient.PostAsJsonAsync("/api/v1/risks/calculate", payload);
            if (!response.IsSuccessStatusCode)
            {
                response = await _httpClient.PostAsJsonAsync("/risks/calculate", payload);
            }

            if (response.IsSuccessStatusCode)
            {
                backendData = await response.Content.ReadFromJsonAsync<JsonObject>();
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning("Failed calling backend risk API, performing fallback: {Message}", ex.Message);
        }

        // Extract or compute score
        double riskScore = backendData?["risk_score"]?.GetValue<double>() ?? ComputeFallbackRiskScore(req);
        string category = backendData?["risk_category"]?.GetValue<string>() ?? GetCategoryFromScore(riskScore);
        string evalId = backendData?["risk_evaluation_id"]?.GetValue<string>() ?? $"EVAL-{Guid.NewGuid().ToString()[..8].ToUpper()}";
        string calculatedAt = backendData?["calculated_at"]?.GetValue<string>() ?? DateTime.UtcNow.ToString("o");

        string color = category.ToUpper() switch
        {
            "CRITICAL" => "#ef4444",
            "VERY_HIGH" => "#f97316",
            "HIGH" => "#eab308",
            "MODERATE" => "#3b82f6",
            _ => "#10b981"
        };

        string recommendation = category.ToUpper() switch
        {
            "CRITICAL" or "VERY_HIGH" => "Immediate field dispatch and structural reinforcement required.",
            "HIGH" => "Schedule maintenance work within 30 days. Priority queue escalated.",
            "MODERATE" => "Monitor during next routine cycle. Routine inspection recommended.",
            _ => "Asset operating within normal safety limits. No immediate action."
        };

        var factorScores = new Dictionary<string, double>
        {
            ["AHS"] = backendData?["ahs"]?.GetValue<double>() ?? Math.Min(100, req.AgeYears * 1.5 + (100 - req.RemainingFiberPct)),
            ["CES"] = backendData?["ces"]?.GetValue<double>() ?? Math.Min(100, req.PopulationExposure * 0.5 + req.ServiceImpact * 0.5),
            ["CQS"] = backendData?["cqs"]?.GetValue<double>() ?? Math.Min(100, (req.FailureCost / 10000.0) + (req.CustomerCount / 100.0)),
            ["DHM"] = backendData?["dhm"]?.GetValue<double>() ?? Math.Min(100, req.WindSpeedMph * 1.2 + (req.FireWeatherIndex ?? 50) * 0.4),
            ["OPS"] = backendData?["ops"]?.GetValue<double>() ?? Math.Min(100, req.DaysOverdue * 1.1),
            ["VES"] = backendData?["ves"]?.GetValue<double>() ?? 15.0
        };

        var drivers = new List<object>();
        if (backendData?["explanation"]?["factors"] is JsonArray arr)
        {
            foreach (var node in arr)
            {
                if (node != null) drivers.Add(node);
            }
        }
        else if (backendData?["drivers"] is JsonArray driversArr)
        {
            foreach (var node in driversArr)
            {
                if (node != null) drivers.Add(node);
            }
        }

        if (drivers.Count == 0)
        {
            drivers.Add(new { name = "Asset Age & Decay", impact = "+32.5%", category = "AHS" });
            drivers.Add(new { name = "Wildfire Weather Index", impact = "+28.0%", category = "DHM" });
            drivers.Add(new { name = "Overdue Maintenance", impact = "+18.4%", category = "OPS" });
        }

        return new BffEnrichedRiskResponse
        {
            RiskEvaluationId = evalId,
            AssetId = req.AssetId,
            TenantId = req.TenantId,
            RiskScore = Math.Round(riskScore, 2),
            RiskCategory = category,
            CategoryColor = color,
            ActionRecommendation = recommendation,
            Confidence = 0.94,
            CalculatedAt = calculatedAt,
            FactorScores = factorScores,
            Drivers = drivers,
            Warnings = req.DaysOverdue > 30 ? new List<string> { $"Asset inspection is {req.DaysOverdue} days overdue!" } : new List<string>()
        };
    }

    public async Task<BffEnrichedFinancialResponse> EvaluateFinancialsAsync(BffFinancialEngineRequest req)
    {
        var payload = new
        {
            asset_id = req.AssetId,
            tenant_id = req.TenantId,
            condition_band = req.ConditionBand,
            replacement_cost = req.ReplacementCost,
            customers = req.Customers,
            outage_hours = req.OutageHours,
            outage_cost_per_customer_hour = req.OutageCostPerCustomerHour,
            wildfire_liability_exposure = req.WildfireLiabilityExposure,
            hftd_tier = req.HftdTier,
            action_cost = req.ActionCost,
            remaining_life = req.RemainingLife,
            design_life = req.DesignLife,
            discount_rate = req.DiscountRate,
            degradation_rate = req.DegradationRate
        };

        JsonObject? backendData = null;
        try
        {
            var response = await _httpClient.PostAsJsonAsync("/api/v1/financial-engine/evaluate", payload);
            if (response.IsSuccessStatusCode)
            {
                backendData = await response.Content.ReadFromJsonAsync<JsonObject>();
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning("Failed calling backend financial engine API: {Message}", ex.Message);
        }

        double eal = backendData?["base_summary"]?["eal"]?.GetValue<double>() ?? (req.ReplacementCost * 0.18 + req.Customers * req.OutageHours * req.OutageCostPerCustomerHour * 0.05);
        double netBenefit = backendData?["base_summary"]?["net_benefit"]?.GetValue<double>() ?? (eal * 3.2 - req.ActionCost);
        double roi = backendData?["base_summary"]?["roi"]?.GetValue<double>() ?? (req.ActionCost > 0 ? netBenefit / req.ActionCost : 2.5);

        string fmtNetBenefit = $"${netBenefit:N2}";
        string fmtRoi = $"{roi * 100:F1}%";
        string fmtEal = $"${eal:N2}";

        string investRec = roi switch
        {
            > 2.0 => "Strong positive return on investment. Immediate capital allocation justified.",
            > 0.0 => "Favorable net benefit. Recommended for upcoming budget allocation.",
            _ => "Negative net benefit. Defer major overhaul unless required by risk threshold."
        };

        return new BffEnrichedFinancialResponse
        {
            EvaluationId = backendData?["evaluation_id"]?.GetValue<string>() ?? $"FIN-{Guid.NewGuid().ToString()[..8].ToUpper()}",
            AssetId = req.AssetId,
            TenantId = req.TenantId,
            CalculationVersion = "1.0",
            Scenarios = (object?)backendData?["scenarios"] ?? new
            {
                do_nothing_eal = eal * 1.5,
                preventative_maint_cost = req.ActionCost,
                full_replacement_cost = req.ReplacementCost
            },
            BaseSummary = (object?)backendData?["base_summary"] ?? new
            {
                eal = Math.Round(eal, 2),
                net_benefit = Math.Round(netBenefit, 2),
                roi = Math.Round(roi, 4)
            },
            InvestmentRecommendation = investRec,
            FormattedNetBenefit = fmtNetBenefit,
            FormattedRoi = fmtRoi,
            FormattedEal = fmtEal
        };
    }

    public async Task<BffEnrichedPriorityResponse> CalculatePolePriorityAsync(BffPolePriorityRequest req)
    {
        var payload = new
        {
            asset_id = req.AssetId,
            tenant_id = req.TenantId,
            composite_risk_score = req.CompositeRiskScore,
            consequence_score = req.ConsequenceScore,
            hazard_exposure_score = req.HazardExposureScore,
            days_overdue = req.DaysOverdue,
            inspection_confidence = req.InspectionConfidence
        };

        JsonObject? backendData = null;
        try
        {
            var response = await _httpClient.PostAsJsonAsync("/api/v1/risks/pole-priority", payload);
            if (response.IsSuccessStatusCode)
            {
                backendData = await response.Content.ReadFromJsonAsync<JsonObject>();
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning("Failed calling backend pole priority API: {Message}", ex.Message);
        }

        double priorityScore = backendData?["priority_score"]?.GetValue<double>() ?? 
            Math.Min(100.0, req.CompositeRiskScore * 0.5 + req.ConsequenceScore * 0.3 + req.DaysOverdue * 0.2);

        string tier;
        string color;
        string urgency;

        if (priorityScore >= 80.0)
        {
            tier = "P1 - EMERGENCY / IMMEDIATE";
            color = "#ef4444";
            urgency = "Dispatch field crew immediately. High public safety risk.";
        }
        else if (priorityScore >= 60.0)
        {
            tier = "P2 - HIGH URGENCY";
            color = "#f97316";
            urgency = "Queue for work order authorization within 7 days.";
        }
        else if (priorityScore >= 40.0)
        {
            tier = "P3 - MEDIUM PRIORITY";
            color = "#eab308";
            urgency = "Schedule replacement in current quarterly plan.";
        }
        else
        {
            tier = "P4 - ROUTINE / DEFERRABLE";
            color = "#10b981";
            urgency = "No work order required at present.";
        }

        return new BffEnrichedPriorityResponse
        {
            AssetId = req.AssetId,
            TenantId = req.TenantId,
            PriorityScore = Math.Round(priorityScore, 1),
            PriorityTier = tier,
            PriorityColor = color,
            UrgencyRecommendation = urgency,
            ScoreBreakdown = new Dictionary<string, object>
            {
                ["composite_risk_component"] = Math.Round(req.CompositeRiskScore * 0.5, 2),
                ["consequence_component"] = Math.Round(req.ConsequenceScore * 0.3, 2),
                ["overdue_component"] = Math.Round(req.DaysOverdue * 0.2, 2)
            }
        };
    }

    public async Task<BffUnifiedAuditResponse> RunUnifiedAuditAsync(string assetId, string tenantId)
    {
        var riskReq = new BffRiskCalculationRequest { AssetId = assetId, TenantId = tenantId };
        var finReq = new BffFinancialEngineRequest { AssetId = assetId, TenantId = tenantId };

        var riskRes = await CalculateRiskAsync(riskReq);
        var finRes = await EvaluateFinancialsAsync(finReq);

        var prioReq = new BffPolePriorityRequest
        {
            AssetId = assetId,
            TenantId = tenantId,
            CompositeRiskScore = riskRes.RiskScore,
            ConsequenceScore = riskRes.FactorScores.GetValueOrDefault("CES", 50.0),
            HazardExposureScore = riskRes.FactorScores.GetValueOrDefault("DHM", 50.0)
        };
        var prioRes = await CalculatePolePriorityAsync(prioReq);

        int historyCount = 1;
        try
        {
            var histResp = await _httpClient.GetAsync($"/api/v1/risks/history/{assetId}?tenant_id={tenantId}");
            if (histResp.IsSuccessStatusCode)
            {
                var list = await histResp.Content.ReadFromJsonAsync<List<JsonObject>>();
                if (list != null) historyCount = list.Count;
            }
        }
        catch (Exception ex)
        {
            _logger.LogWarning("Could not fetch history for asset {AssetId}: {Message}", assetId, ex.Message);
        }

        string overallHealthStatus = "STABLE";
        if (riskRes.RiskCategory is "CRITICAL" or "VERY_HIGH" || prioRes.PriorityScore >= 80)
        {
            overallHealthStatus = "CRITICAL ACTION REQUIRED";
        }
        else if (riskRes.RiskCategory == "HIGH")
        {
            overallHealthStatus = "ELEVATED RISK";
        }

        return new BffUnifiedAuditResponse
        {
            AssetId = assetId,
            TenantId = tenantId,
            Timestamp = DateTime.UtcNow.ToString("o"),
            Risk = riskRes,
            Financials = finRes,
            Priority = prioRes,
            HistoryCount = historyCount,
            OverallHealthStatus = overallHealthStatus
        };
    }

    private static double ComputeFallbackRiskScore(BffRiskCalculationRequest req)
    {
        double score = (req.AgeYears * 0.8) + (100 - req.RemainingFiberPct) * 0.4 + (req.LeanDeg * 4.0) + (req.DaysOverdue * 0.3);
        return Math.Min(100.0, Math.Max(0.0, score));
    }

    private static string GetCategoryFromScore(double score) => score switch
    {
        >= 85.0 => "CRITICAL",
        >= 70.0 => "VERY_HIGH",
        >= 50.0 => "HIGH",
        >= 30.0 => "MODERATE",
        _ => "LOW"
    };
}
